File: bindings/go/examples/basic_example.go
package main

import (
	"fmt"
	"net/netip"

	"github.com/MuriloChianfa/liblpm/go/liblpm"
)

func main() {
	// Create an IPv4 routing table
	table, err := liblpm.NewTableIPv4()
	if err != nil {
		panic(err)
	}
	defer table.Close()

	// Add some routes
	routes := []struct {
		prefix  string
		nextHop liblpm.NextHop
		name    string
	}{
		{"10.0.0.0/8", 100, "Private Network A"},
		{"192.168.0.0/16", 200, "Private Network C"},
		{"172.16.0.0/12", 300, "Private Network B"},
		{"8.8.8.0/24", 400, "Google DNS"},
	}

	fmt.Println("Adding routes to table:")
	for _, route := range routes {
		prefix := netip.MustParsePrefix(route.prefix)
		err := table.Insert(prefix, route.nextHop)
		if err != nil {
			fmt.Printf("  Error adding %s: %v\n", route.prefix, err)
		} else {
			fmt.Printf("  Added: %s -> next hop %d (%s)\n", route.prefix, route.nextHop, route.name)
		}
	}

	// Perform some lookups
	fmt.Println("\nPerforming lookups:")
	lookupAddrs := []string{
		"10.1.1.1",
		"192.168.1.1",
		"172.16.5.10",
		"8.8.8.8",
		"1.1.1.1", // No match
	}

	for _, addrStr := range lookupAddrs {
		addr := netip.MustParseAddr(addrStr)
		nextHop, found := table.Lookup(addr)
		
		if found {
			// Find the route name
			routeName := "Unknown"
			for _, route := range routes {
				if uint32(route.nextHop) == uint32(nextHop) {
					routeName = route.name
					break
				}
			}
			fmt.Printf("  %s -> next hop %d (%s)\n", addrStr, nextHop, routeName)
		} else {
			fmt.Printf("  %s -> No route found\n", addrStr)
		}
	}

	// Demonstrate batch lookup
	fmt.Println("\nBatch lookup example:")
	batchAddrs := []netip.Addr{
		netip.MustParseAddr("10.1.1.1"),
		netip.MustParseAddr("192.168.1.1"),
		netip.MustParseAddr("172.16.5.10"),
		netip.MustParseAddr("1.1.1.1"),
	}

	results, err := table.LookupBatch(batchAddrs)
	if err != nil {
		panic(err)
	}

	for i, addr := range batchAddrs {
		if results[i].IsValid() {
			fmt.Printf("  %s -> next hop %d\n", addr, results[i])
		} else {
			fmt.Printf("  %s -> No route\n", addr)
		}
	}

	// IPv6 example
	fmt.Println("\n--- IPv6 Example ---")
	tableV6, err := liblpm.NewTableIPv6()
	if err != nil {
		panic(err)
	}
	defer tableV6.Close()

	// Add IPv6 routes
	ipv6Routes := []struct {
		prefix  string
		nextHop liblpm.NextHop
	}{
		{"2001:db8::/32", 1000},
		{"2001:db9::/32", 2000},
		{"fc00::/7", 3000}, // ULA
	}

	fmt.Println("Adding IPv6 routes:")
	for _, route := range ipv6Routes {
		prefix := netip.MustParsePrefix(route.prefix)
		err := tableV6.Insert(prefix, route.nextHop)
		if err != nil {
			fmt.Printf("  Error adding %s: %v\n", route.prefix, err)
		} else {
			fmt.Printf("  Added: %s -> next hop %d\n", route.prefix, route.nextHop)
		}
	}

	// IPv6 lookups
	fmt.Println("\nIPv6 lookups:")
	ipv6Addrs := []string{
		"2001:db8::1",
		"2001:db9::1",
		"fc00::1",
		"2001:dba::1", // No match
	}

	for _, addrStr := range ipv6Addrs {
		addr := netip.MustParseAddr(addrStr)
		nextHop, found := tableV6.Lookup(addr)
		
		if found {
			fmt.Printf("  %s -> next hop %d\n", addrStr, nextHop)
		} else {
			fmt.Printf("  %s -> No route found\n", addrStr)
		}
	}
}

File: bindings/python/examples/basic_example.py
#!/usr/bin/env python3
"""Basic example demonstrating liblpm Python bindings.

This example shows the fundamental operations: creating a table,
inserting routes, looking up addresses, and cleaning up resources.
"""

from ipaddress import IPv4Address, IPv4Network, IPv6Address, IPv6Network

from liblpm import LpmTableIPv4, LpmTableIPv6


def ipv4_example():
    """Demonstrate IPv4 routing table operations."""
    print("=" * 60)
    print("IPv4 Routing Table Example")
    print("=" * 60)
    
    # Create an IPv4 routing table using context manager
    with LpmTableIPv4() as table:
        print(f"\nCreated IPv4 table with algorithm: {table.algorithm}")
        
        # Insert some routes
        routes = [
            ('0.0.0.0/0', 1),         # Default route
            ('10.0.0.0/8', 10),        # Private network
            ('192.168.0.0/16', 168),   # Private network
            ('192.168.1.0/24', 1681),  # More specific
            ('8.8.8.0/24', 888),       # Google DNS
        ]
        
        print("\nInserting routes:")
        for prefix, next_hop in routes:
            table.insert(prefix, next_hop)
            print(f"  {prefix} -> {next_hop}")
        
        print(f"\nTable now has {table.num_prefixes} prefixes")
        
        # Look up some addresses
        addresses = [
            '192.168.1.100',  # Should match /24 (most specific)
            '192.168.2.100',  # Should match /16
            '10.0.0.1',       # Should match /8
            '8.8.8.8',        # Should match Google DNS
            '1.1.1.1',        # Should match default route
        ]
        
        print("\nLooking up addresses:")
        for addr_str in addresses:
            addr = IPv4Address(addr_str)
            next_hop = table.lookup(addr)
            print(f"  {addr_str} -> next_hop: {next_hop}")
    
    print("\nTable automatically closed after 'with' block")


def ipv6_example():
    """Demonstrate IPv6 routing table operations."""
    print("\n" + "=" * 60)
    print("IPv6 Routing Table Example")
    print("=" * 60)
    
    with LpmTableIPv6() as table:
        print(f"\nCreated IPv6 table with algorithm: {table.algorithm}")
        
        # Insert some routes
        routes = [
            ('::/0', 1),                    # Default route
            ('2001:db8::/32', 2001),         # Documentation prefix
            ('2001:db8:1::/48', 20011),      # More specific
            ('fd00::/8', 0xfd),              # Unique local
            ('fe80::/10', 0xfe80),           # Link-local
        ]
        
        print("\nInserting routes:")
        for prefix, next_hop in routes:
            table.insert(prefix, next_hop)
            print(f"  {prefix} -> {next_hop}")
        
        print(f"\nTable now has {table.num_prefixes} prefixes")
        
        # Look up some addresses
        addresses = [
            '2001:db8:1::1',    # Should match /48 (most specific)
            '2001:db8:ffff::1', # Should match /32
            'fd00::1',          # Should match unique local
            'fe80::1',          # Should match link-local
            '2607:f8b0::1',     # Should match default route
        ]
        
        print("\nLooking up addresses:")
        for addr_str in addresses:
            addr = IPv6Address(addr_str)
            next_hop = table.lookup(addr)
            print(f"  {addr_str} -> next_hop: {next_hop}")
    
    print("\nTable automatically closed after 'with' block")


def main():
    """Run all examples."""
    print("liblpm Python Bindings - Basic Example")
    print("======================================\n")
    
    ipv4_example()
    ipv6_example()
    
    print("\n" + "=" * 60)
    print("All examples completed successfully!")
    print("=" * 60)


if __name__ == '__main__':
    main()
